### ISSUE TITLE
Support pyQuil parametric compilation 

### ISSUE THREAD

**Issue by Misty-W** [unitaryfund/mitiq#834]
Parametric compilation saves time by compiling the program ansatz once, instead of compiling each time the parameters are updated. In pyQuil 2.28.1, a `memory_map` argument is used to substitute in values for previously-declared Quil variables in the pre-compiled executable. It would be helpful to be able to leverage the runtime improvement from parametric compilation when using Mitiq. Right now using a pyQuil memory map with Mitiq results in an error (details below).

The following code snippet (no mitigation) runs through and gives the expected result:

    from pyquil import get_qc, Program
    from pyquil.gates import RY, MEASURE
    import mitiq
    from mitiq import zne
    from mitiq.zne import execute_with_zne
    import numpy as np
    from scipy import optimize
    from functools import partial
    # Initialize the quantum device
    qc = get_qc("2q-noisy-qvm")

    # Set up quantum circuit
    def ansatz():
        program = Program()
        theta = program.declare('theta', memory_type='REAL')
        ro = program.declare('ro', memory_type='BIT', memory_size=1)
        program += RY(theta, 0)
        program += MEASURE(0, ro[0])
        return program

    print(ansatz())

    def expectation_fn(theta, executable):
        bitstrings = qc.run(executable, memory_map= {'theta': [theta]})
        return np.mean(bitstrings[:,0])

    def expectation(thetas, program):
        program.wrap_in_numshots_loop(1000) 
        executable = qc.compile(program)
        bitstrings = qc.run(executable, memory_map={'theta': thetas})
        result = np.mean(bitstrings[:, 0])
        return -result

    thetas = np.linspace(0, 2*np.pi, 21)
    program = ansatz()
    results = []
    program.wrap_in_numshots_loop(1000) 
    quil_prog = qc.compiler.quil_to_native_quil(program, protoquil=True)

    for theta in thetas:
        results.append(expectation_fn(theta, quil_prog))`

Adding the code snippet below results in an error:

    program = ansatz()
    results_zne = []
    for theta in thetas:
        results_zne.append(execute_with_zne(quil_prog, partial(expectation_fn, theta)))

The error message is 

> QVMError: At line 28: Bad memory region name "theta" in MOVE. This is probably due to either:
    * a missing DECLARE for this memory,
    * a misspelling of the memory reference, or
    * a misspelling of the DECLAREd memory.

On the surface it looks like a problem with Quil, but the error only occurs with the addition of `execute_w_zne`. I suspect the error arises because Mitiq does not support `memory_map`, hence I filed the issue against Mitiq.

**Comment by rmlarose** [comment]
Thanks @Misty-W. I haven't looked into it yet but I suspect that the pyQuil program -> Cirq circuit -> pyQuil program conversion causes the "declare" memory reference to be lost.

Some of these more detailed features in individual circuits are harder to maintain during conversions. We should definitely support this if it is possible though.

**Comment by rmlarose** [comment]
Conversions indeed drop the memory reference:

```python
import pyquil
from mitiq import zne

program = pyquil.Program()
theta = program.declare('theta', memory_type='REAL')
program += pyquil.gates.RY(theta, 0)
print(program)
# DECLARE theta REAL[1]
# RY(theta) 0

scaled = zne.scaling.fold_gates_at_random(program, 1)
print(scaled)
# RY(theta[0]/pi*pi) 0
```
Working on a fix for this.

**Comment by rmlarose** [comment]
#835 fixes this. You'll also need to use something like [`mitiq.interface.mitiq_pyquil.pyquil_utils.generate_qcs_executor`](https://github.com/unitaryfund/mitiq/blob/master/mitiq/interface/mitiq_pyquil/pyquil_utils.py#L27) to make sure the memory declarations are correct: The scaled circuits will have names like "m0" but you need a special one called "ro" to work in PyQuil.

**Comment by Misty-W** [comment]
Thanks @rmlarose, appreciate the quick fix! 

As for declaring the "ro", right now I have

    def ansatz():
        program = Program()
        theta = program.declare('theta', memory_type='REAL')
        ro = program.declare('ro', memory_type='BIT', memory_size=1)
        program += RY(theta, 0)
        program += MEASURE(0, ro[0])
        return program

    def expectation_fn(theta, executable):
        bitstrings = qc.run(executable, memory_map= {'theta': [theta]})
        return np.mean(bitstrings[:,0])

    program = ansatz()
    program.wrap_in_numshots_loop(1000) 
    quil_prog = qc.compiler.quil_to_native_quil(program, protoquil=True)
    results_zne = []
    thetas = np.linspace(0, 2*np.pi, 21) 
    for theta in thetas:
        results_zne.append(execute_with_zne(quil_prog, partial(expectation_fn, theta)))

**Comment by rmlarose** [comment]
Hi Misty, I also fixed the other issues you were running into in #835. Here's your (slightly modified) example which I can successfully run on Mitiq/master now:

```python
import numpy as np

from pyquil import get_qc, Program
from pyquil.gates import RY, MEASURE
from mitiq import zne

# Initialize the quantum device
qc = get_qc("2q-noisy-qvm")

# Set up quantum circuit
def ansatz():
    program = Program()
    theta = program.declare("theta", memory_type="REAL")
    ro = program.declare("ro", memory_type="BIT", memory_size=1)
    program += RY(theta, 0)
    program += MEASURE(0, ro[0])
    return program

def expectation_fn(executable: Program, theta) -> float:
    bitstrings = qc.run(executable, memory_map={"theta": [theta]})
    return np.mean(bitstrings[:, 0])

program = ansatz()
program.wrap_in_numshots_loop(1000)
quil_prog = qc.compiler.quil_to_native_quil(program, protoquil=True)

thetas = np.linspace(0, 2 * np.pi, 2)
results = []
for theta in thetas:
    results.append(expectation_fn(quil_prog, theta))
print(results)

results_zne = []
for theta in thetas:
    results_zne.append(
        zne.execute_with_zne(quil_prog, lambda p: expectation_fn(p, theta))
    )
print(results_zne)
```

(The story with `"ro"` is that (i) you were doing it correctly but (ii) when `execute_with_zne` was called, the name of the register in the scaled circuits would get changed to something that wasn't `"ro"`, so the measurements would be empty. But this is fixed now.)

**Comment by Misty-W** [comment]
@rmlarose, thanks again for the fix and the explanations! I tested the "slightly edited example" in my environment last night, and unfortunately I'm still getting the same error. 

> QVMError: At line 28: Bad memory region name "theta" in MOVE. This is probably due to either:
>  * a missing DECLARE for this memory,
>  * a misspelling of the memory reference, or
>  * a misspelling of the DECLAREd memory.

I made sure to update Mitiq from [https://github.com/unitaryfund/mitiq.git](https://github.com/unitaryfund/mitiq.git) before running my example. I also looked at the conversions.py file in my environment directory. I can see it is updated with your commits from the last 2 days, so I believe I have the correct version of Mitiq installed.

Next step is to try a fresh environment, just in case. Any ideas are welcome!
